src/openlcb/SimpleStack.hxx

```diff
@@ -358,40 +258,216 @@ protected:
-        config_num_datagram_registry_entries(), config_num_datagram_clients()};
-    MemoryConfigHandler memoryConfigHandler_{
-        &datagramService_, nullptr, config_num_memory_spaces()};
+    MemoryConfigHandler memoryConfigHandler_ {
+        datagramService_, nullptr, config_num_memory_spaces()};
 
     /// All packets are forwarded to this hub in gridconnect format, if
     /// needed. Will be initialized upon first use.
     std::unique_ptr<HubFlow> gcHub_;
     /// Bridge between canHub_ and gcHub_. Lazily initialized.
     std::unique_ptr<GCAdapterBase> gcAdapter_;
-
     /// Stores and keeps ownership of optional components.
     std::vector<std::unique_ptr<Destructable>> additionalComponents_;
 };
 
-/// CAN-based stack with DefaultNode.
+/// SimpleStack with a CAN-bus based interface and IO functions for CAN-bus.
+class SimpleCanStackBase : public SimpleStackBase
+{
+public:
+    SimpleCanStackBase(const openlcb::NodeID node_id);
+
+    /// @returns the CanHubFlow to which this stack is talking to. This hub
+    /// flow usually has two members: the interface object from the software
+    /// stack and the hardware connection via which to connect to the physical
+    /// bus (which may be a device driver or a gridconnect protocol converter).
+    CanHubFlow *can_hub()
+    {
+        return &static_cast<CanPhysicalIf *>(ifaceHolder_.get())->canHub0_;
+    }
+
+    /// Adds a CAN bus port with synchronous driver API.
+    void add_can_port_blocking(const char *device)
+    {
+        int can_fd = ::open(device, O_RDWR);
+        HASSERT(can_fd >= 0);
+        auto *port = new FdHubPort<CanHubFlow>(
+            can_hub(), can_fd, EmptyNotifiable::DefaultInstance());
+        additionalComponents_.emplace_back(port);
+    }
+
+#ifdef __FreeRTOS__
+    /// Adds a CAN bus port with asynchronous driver API.
+    ///
+    /// @deprecated: most current FreeRTOS drivers use the the select-based
+    /// asynchronous API, so they need add_can_port_select().
+    void add_can_port_async(const char *device)
+    {
+        auto *port = new HubDeviceNonBlock<CanHubFlow>(can_hub(), device);
+        additionalComponents_.emplace_back(port);
+    }
+
+    /// Adds a CAN bus port with select-based asynchronous driver API.
+    void add_can_port_select(const char *device)
+    {
+        auto *port = new HubDeviceSelect<CanHubFlow>(can_hub(), device);
+        additionalComponents_.emplace_back(port);
+    }
+
+    /// Adds a CAN bus port with select-based asynchronous driver API.
+    /// @param fd file descriptor to add to can hub
+    /// @param on_error Notifiable to wakeup on error
+    void add_can_port_select(int fd, Notifiable *on_error = nullptr)
+    {
+        auto *port = new HubDeviceSelect<CanHubFlow>(can_hub(), fd, on_error);
+        additionalComponents_.emplace_back(port);
+    }
+#endif
+
+    /// Adds a gridconnect port to the CAN bus.
+    void add_gridconnect_port(const char *path, Notifiable *on_exit = nullptr);
```

Comment:
Possibly same on this method.

Reply:
Also is there much difference between this and the method below?

Reply:
on linux you want to use the TTY version because it uses the terminal kernel API to set the baud rate and most importantly disable echo.

Reply:
Ok that makes sense.